drivers/usb/usbccgp/function.c

```diff
@@ -654,24 +654,61 @@ USBCCGP_LegacyEnum(
             return STATUS_UNSUCCESSFUL;
         }
 
-        //
-        // init function descriptor
-        //
-        FDODeviceExtension->FunctionDescriptor[Index].FunctionNumber = Index;
-        FDODeviceExtension->FunctionDescriptor[Index].NumberOfInterfaces = 1;
-        FDODeviceExtension->FunctionDescriptor[Index].InterfaceDescriptorList = AllocateItem(NonPagedPool, sizeof(PUSB_INTERFACE_DESCRIPTOR) * 1);
-        if (!FDODeviceExtension->FunctionDescriptor[Index].InterfaceDescriptorList)
+        SubIndex = 0;
+        if (InterfaceDescriptor->bInterfaceClass == USB_DEVICE_CLASS_AUDIO)
         {
+            // AUDIO CLASS lets group all audio interfaces together
+            //
+            // init function descriptor
+            //
+            FDODeviceExtension->FunctionDescriptor[Index].FunctionNumber = Index;
+            FDODeviceExtension->FunctionDescriptor[Index].NumberOfInterfaces = 1;
+            FDODeviceExtension->FunctionDescriptor[Index].InterfaceDescriptorList = AllocateItem(NonPagedPool, sizeof(PUSB_INTERFACE_DESCRIPTOR) * FDODeviceExtension->ConfigurationDescriptor->bNumInterfaces);
+            if (!FDODeviceExtension->FunctionDescriptor[Index].InterfaceDescriptorList)
+            {
+                //
+                // no memory
+                //
+                return STATUS_INSUFFICIENT_RESOURCES;
+            }
+
             //
-            // no memory
+            // store interface descriptor
             //
-            return STATUS_INSUFFICIENT_RESOURCES;
+            FDODeviceExtension->FunctionDescriptor[Index].InterfaceDescriptorList[SubIndex] = InterfaceDescriptor;
+            do
+            {
+                NextInterfaceDescriptor = USBD_ParseConfigurationDescriptorEx(FDODeviceExtension->ConfigurationDescriptor, FDODeviceExtension->ConfigurationDescriptor, Index + SubIndex + 1, 0, -1, -1, -1);
+                if (!NextInterfaceDescriptor || NextInterfaceDescriptor->bInterfaceClass != USB_DEVICE_CLASS_AUDIO)
+                {
+                    break;
+                }
+                SubIndex++;
```

Comment:
Should ASSERT that this is within bound of `InterfaceDescriptorList`.

Reply:
done